### ISSUE TITLE
Introduce `down` as opposite of `up`

### ISSUE THREAD

**Issue by oxinabox** [JuliaLang/Pkg.jl#1062]
I wish to test that I have set the lower bounds in my project.toml correctly.
I thus want to check that with the oldest package versions that can be compatibly installed,
my tests still pass.

So I want to do the oppoite of `]up`

**Comment by fredrikekre** [comment]
This should probably just be flag to `pkg> test` rather than a `pkg> down` command.

**Comment by oxinabox** [comment]
No, because when debugging I want to instantiate the set of package versions I am testing on.
And then I can incrementally raise my lower bound, and run `pkg> resolve` and then test again.

**Comment by StefanKarpinski** [comment]
`pkg> down` seems reasonable to me although it might just be a shortcut for `pkg> up --lower` or something like that.

**Comment by cossio** [comment]
Related post from rust: https://users.rust-lang.org/t/psa-please-specify-precise-dependency-versions-in-cargo-toml/71277.

**Comment by marius311** [comment]
Are there any hacks to do this right now? I agree it'd be really useful for exactly the reason stated by OP. The place I'd really want to use it is for my CI; if available, I'd test with both the highest and lowest possible compatible dependency versions.

**Comment by KristofferC** [comment]
@carlobaldassi, what would be required in the resolver to support this? I am guessing just redefining some stuff in https://github.com/JuliaLang/Pkg.jl/blob/master/src/Resolve/versionweights.jl but are there more things? If you could have some example code I could integrate it easily.

**Comment by carlobaldassi** [comment]
@KristofferC I think that the main change would be to reverse the sorting order of versions of the required packages (while leaving the rest unchanged) here: https://github.com/JuliaLang/Pkg.jl/blob/76070d295fc4a1f27f852e05400bbc956962e084/src/Resolve/graphtype.jl#L128

Then I think it might be sufficient to negate the second argument `vweight[p0][v0]` here: https://github.com/JuliaLang/Pkg.jl/blob/76070d295fc4a1f27f852e05400bbc956962e084/src/Resolve/maxsum.jl#L61

I suspect this may be all (plus of course passing around some flags), but it's hard to tell without testing. Unfortunately, due to personal matters, I don't have the time for doing that myself at the moment.

On the other hand, I wonder whether it would be best to simply call the solver with the versions of the explicitly required dependencies fixed to their declared lower bounds. In my mind, that would be simpler, safer, and more precise.
There would be no need to further increase the complexity of the resolver code, and it would actually test that the declared lower bounds work. After all, if the resolver finds that it cannot set a package version to its declared lower bound due to an incompatibility with some other required package, then the lower bound should be raised, no? Or am I missing something?

**Comment by KristofferC** [comment]
Thanks! I will experiment with those changes.

>  After all, if the resolver finds that it cannot set a package version to its declared lower bound due to an incompatibility with some other required package, then the lower bound should be raised, no? Or am I missing something?

It's definitely worth trying, I will try that too and see what happens.

**Comment by KristofferC** [comment]
> In short this seems no more sensible than making `] up` fix the versions to the upper bounds declared, for basically the same reasons.

I am not sure about that because the reason you have a lower bound is that you require something specific and known that happened in that version, while for an upper bound you are protecting yourself against future unknown breakage. So it isn't self evident to me that these are completely symmetric things and that `down` should be handled completely equal to `up` except with version checking inverted.

For example, even with `down` I think you still want the highest version possible of transitive dependencies because you cannot change `compat` to directly influence which of those you are going to get. So I like the suggestion that `down` would only try to downgrade direct dependencies while leaving transitive dependencies to still be maximized.

**Comment by KristofferC** [comment]
The reason for having `down` is specifically to test that your lower bounds are correct and it feels important to actually hit exactly those lower bounds in `down`. If all you are testing is lower versions of your package (but these are still higher than your lower bounds), then it feels like the point of doing this at all is more or less lost.

**Comment by oxinabox** [comment]
> The reason for having down is specifically to test that your lower bounds are correct and it feels important to actually hit exactly those lower bounds in down.

I would argue the point is to pick somewhere on the downward parento-fountier of achievable (not declared) downness.
Ideally would check every point on the pareto-frontier of downness, but that is potentially too much work.
